File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int answer = 0;
        
        int[] dx = {1, 0, -1, 0};
        int[] dy = {0, 1, 0, -1};
        boolean[][] visited = new boolean[maps.length][maps[0].length];
        
        Queue<Position> queue = new LinkedList<>();
        queue.add(new Position(0,0,1));
        visited[0][0] = true;
        
        while(!queue.isEmpty()) {
            Position position = queue.poll();
            
            if (position.x == maps.length - 1 && position.y == maps[0].length - 1) {
                answer = position.moveCnt;
                break;
            }
            for (int i = 0; i < 4; i++) {
                int nx = position.x + dx[i];
                int ny = position.y + dy[i];
                
                if (nx < 0 || nx >= maps.length || ny < 0 || ny >= maps[0].length) {
                   continue;
                }
                
                if (!visited[nx][ny] && maps[nx][ny] == 1) {
                    visited[nx][ny] = true;
                    queue.add(new Position(nx, ny, position.moveCnt + 1));
                }
            }
        }
        
        if (answer == 0) {
            answer = -1;
        }
        return answer;
    }
    
    class Position {
        
        int x;
        int y;
        int moveCnt;
        
        public Position (int x, int y, int moveCnt) {
            this.x = x;
            this.y = y;
            this.moveCnt = moveCnt;
        }
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = bfs(maps)
    
    
    return answer

# 캐릭터는 항상 좌측 상단 위치에서 시작
def bfs(maps):
    visited = [[0] * len(maps[0]) for i in range(len(maps))]
    
    start_x = 0
    start_y = 0
    
    end_x = len(maps) - 1
    end_y = len(maps[0]) - 1
    
    dx = [0, 1, 0, -1]
    dy = [1, 0, -1, 0]
    
    q = deque()
    visited[start_x][start_y] = 1
    q.append([start_x, start_y, 1])
    
    
    while q:
        p = q.popleft()
        
        p_x = p[0]
        p_y = p[1]
        move = p[2]
        
        if (p_x == end_x and p_y == end_y):
            return move
        
        for i in range(4):
            nx = p_x + dx[i]
            ny = p_y + dy[i]
            
            if nx < 0 or nx >= len(maps) or ny < 0 or ny >= len(maps[0]) or maps[nx][ny] == 0:
                continue
            
            if visited[nx][ny] == 0:
                visited[nx][ny] = 1
                q.append([nx, ny, move + 1])
        
    return -1
    
    
